kubernetes/config/incluster_config.c

```diff
@@ -0,0 +1,154 @@
+        return -1;
+    }
+
+    if (-1 == access(fileName, R_OK)) {
+        fprintf(stderr, "%s: No permission to read the file %s.[%s]\n", fname, fileName, strerror(errno));
+        return -1;
+    }
+
+    return 0;
+}
+
+static int setBasePathInCluster(char **pBasePath)
+{
+    static char fname[] = "setBasePathInCluster()";
+
+    const char *service_host_env = getenv(SERVICE_HOST_ENV_NAME);
+    if (!service_host_env) {
+        fprintf(stderr, "%s: Cannot retrieve the kubernetes service host inside a pod by the env %s.\n", fname, SERVICE_HOST_ENV_NAME);
+        return -1;
+    }
+
+    const char *service_port_env = getenv(SERVICE_PORT_ENV_NAME);
+    if (!service_port_env) {
+        fprintf(stderr, "%s: Cannot retrieve the kubernetes service port inside a pod by the env %s.\n", fname, SERVICE_PORT_ENV_NAME);
+        return -1;
+    }
+
+    char *basePath = calloc(strlen(SERVICE_HTTPS_PREFIX) + strlen(service_host_env) + strlen(service_port_env) + 2 /* 1 for ':', 1 for '\0' */ , sizeof(char));
+    if (!basePath) {
+        fprintf(stderr, "%s: Cannot allocate the memory for base path for kubernetes service.\n", fname);
+        return -1;
+    }
+
+    sprintf(basePath, "%s%s:%s", SERVICE_HTTPS_PREFIX, service_host_env, service_port_env);
+    *pBasePath = basePath;
+    return 0;
+}
+
+static int setSslConfigInCluster(sslConfig_t ** pSslConfig, const char *caFileName)
+{
+    const char *ca_file = NULL;
+    int insecure_skip_tls_verify = 0;
+
+    if (0 == checkServiceAccountFile(caFileName)) {
+        ca_file = caFileName;
+    } else {
+        insecure_skip_tls_verify = 1;
+    }
+
+    sslConfig_t *sc = sslConfig_create(NULL, NULL, ca_file, insecure_skip_tls_verify);
+    if (!sc) {
+        return -1;
+    }
+
+    *pSslConfig = sc;
+    return 0;
+}
+
+static int readTokenInCluster(char *token, int token_buf_size)
+{
+    static char fname[] = "readTokenInCluster()";
+
+    if (0 != checkServiceAccountFile(SERVICE_TOKEN_FILENAME)) {
+        return -1;
+    }
+
+    FILE *fp;
+    fp = fopen(SERVICE_TOKEN_FILENAME, "r");
+    if (NULL == fp) {
+        fprintf(stderr, "%s: Failed to open file %s.[%s]", fname, SERVICE_TOKEN_FILENAME, strerror(errno));
+        return -1;
+    }
+
+    while (fgets(token, token_buf_size, fp) != NULL) {
+        ;
+    }
+
+    fclose(fp);
+    return 0;
+}
+
+static int setApiKeysInCluster(list_t ** pApiKeys)
+{
+    static char fname[] = "setApiKeysInCluster()";
+
+    list_t *apiKeys = list_create();
+    if (!apiKeys) {
+        fprintf(stderr, "%s: Cannot allocate the memory for api kyes for kubernetes service.\n", fname);
+        return -1;
+    }
+
+    char token[BEARER_TOKEN_BUFFER_SIZE];
+    memset(token, 0, sizeof(token));
+    if (0 == readTokenInCluster(token, BEARER_TOKEN_BUFFER_SIZE)) {
+        char tokenValue[BEARER_TOKEN_BUFFER_SIZE];
+        memset(tokenValue, 0, sizeof(tokenValue));
+        sprintf(tokenValue, BEARER_TOKEN_TEMPLATE, token);
```

Comment:
Please use `snprintf` to prevent a buffer overflow.

Reply:
OK.